src/editor/editor_brush_ui.cpp

```diff
@@ -0,0 +1,418 @@
+//      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//      GNU General Public License for more details.
+//
+//      You should have received a copy of the GNU General Public License
+//      along with this program; if not, write to the Free Software
+//      Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+//      02111-1307, USA.
+//
+
+//@{
+
+/*----------------------------------------------------------------------------
+--  Includes
+----------------------------------------------------------------------------*/
+
+
+#include "editor_brush_ui.h"
+
+#include "editor.h"
+
+
+/*----------------------------------------------------------------------------
+--  Defines
+----------------------------------------------------------------------------*/
+
+/*----------------------------------------------------------------------------
+--  Variables
+----------------------------------------------------------------------------*/
+
+/*----------------------------------------------------------------------------
+--  Functions
+----------------------------------------------------------------------------*/
+
+void CSetOfCtrls::show()
+{
+	for(auto &ctrl : hiddenControls) {
+		ctrl->setVisible(true);
+	}
+	hiddenControls.clear();
+}
+
+void CSetOfCtrls::hide()
+{
+	if (!hiddenControls.empty()) {
+		return;
+	}
+	for(auto &ctrl : controls) {
+		if (ctrl->isVisible()) {
+			ctrl->setVisible(false);
+			hiddenControls.push_back(ctrl);
+		}
+	}
+}
+
+Vec2i CSetOfCtrls::bottomRight() const
+{
+	int x = 0;
+	int y = 0;
+	for (const auto ctrl : controls) {
+		x = std::max(x, ctrl->getX() + ctrl->getWidth());
+		y = std::max(y, ctrl->getY() + ctrl->getHeight());
+	}
+	return Vec2i(x, y);
+}
+
+void CBrushControlsUI::reloadBrushes()
+{
+	const auto prevSelectedItem = brushesList->getElementAt(brushSelect->getSelected());
+	const auto newBrushesList = std::make_unique<StringListModel>(Editor.brushes.getBrushesNames());
+	brushesList.reset(newBrushesList.get());
+	brushSelect->setListModel(brushesList.get());
+
+	const std::string currentBrush = Editor.brushes.getCurrentBrush().getName();
+	if (const auto idx = brushesList->getIdxOfElement(currentBrush) != -1) {
+		brushSelect->setSelected(idx);
+	} else {
+		brushSelect->setSelected(0);
+	}
+	reloadCtrlSettings();
+}
+
+void CBrushControlsUI::Init()
+{
+	brushesList = std::make_unique<StringListModel>(Editor.brushes.getBrushesNames());
+	brushSelect = std::make_unique<gcn::DropDown>(brushesList.get());
+	brushSelect->setFocusable(false);
+	brushSelect->setFont(&GetGameFont());
+	brushSelect->setWidth(UIRectangle.width - 10);
+	brushSelect->setBaseColor(baseColor);
+	brushSelect->setForegroundColor(foregroundColor);
+	brushSelect->setBackgroundColor(backgroundColor);
+	parrent->add(brushSelect.get());
+	brushSelect->setPosition(UIRectangle.x + 5, UIRectangle.y + 5);
+
+	controlSets[ECtrlSets::cSelectBrush].add(dynamic_cast<gcn::Widget *>(brushSelect.get()));
```

Comment:
Is the `dynamic_cast` needed?

Reply:
It compiles without dynamic_cast as well, but at least I have that annoying IDE warning:

> argument of type "std::unique_ptr<gcn::DropDown, std::default_delete<gcn::DropDown>>::pointer" (aka "gcn::DropDown *") is incompatible with parameter of type "gcn::Widget *"